File: web/src/shared/protocol/terminal.ts
import { bytesToBase64 } from '@/shared/lib/base64';

export const MAX_TERMINAL_CHUNK_BYTES = 16 * 1024;
const MAX_TERMINAL_BASE64_BYTES = Math.ceil(MAX_TERMINAL_CHUNK_BYTES / 3) * 4;
export const TERMINAL_DIM_MIN = 2;
export const TERMINAL_DIM_MAX = 500;
const MAX_TERMINAL_ID_BYTES = 128;
const TERMINAL_ERROR_CODES = new Set([
  'forbidden',
  'invalid-request',
  'unavailable',
  'delivery-unknown',
  'limit-exceeded',
]);

export interface TerminalOpenedFrame {
  t: 'terminal_opened';
  requestId: string;
  terminalId: string;
  cwd: string;
  cols: number;
  rows: number;
}

export interface TerminalOutputFrame {
  t: 'terminal_output';
  terminalId: string;
  seq: number;
  data: string;
}

export interface TerminalExitFrame {
  t: 'terminal_exit';
  terminalId: string;
  exitCode?: number;
  signal?: string;
}

export interface TerminalErrorFrame {
  t: 'terminal_error';
  requestId?: string;
  terminalId?: string;
  code: 'forbidden' | 'invalid-request' | 'unavailable' | 'delivery-unknown' | 'limit-exceeded';
  message: string;
  retryable: boolean;
}

export type TerminalDownstreamFrame =
  | TerminalOpenedFrame
  | TerminalOutputFrame
  | TerminalExitFrame
  | TerminalErrorFrame;

export const terminalOpen = (
  requestId: string,
  projectId: string,
  cols: number,
  rows: number,
) => ({ t: 'terminal_open', requestId, projectId, cols, rows });

export const terminalInput = (terminalId: string, seq: number, data: string) => ({
  t: 'terminal_input',
  terminalId,
  seq,
  data,
});

export const terminalResize = (terminalId: string, cols: number, rows: number) => ({
  t: 'terminal_resize',
  terminalId,
  cols,
  rows,
});

export const terminalCloseByRequest = (requestId: string) => ({
  t: 'terminal_close',
  requestId,
});

export const terminalClose = (terminalId: string) => ({ t: 'terminal_close', terminalId });

/** xterm `onData` 是 UTF-16 string；TextEncoder 将其还原为 PTY UTF-8 输入字节。 */
export function encodeTerminalInput(data: string): string[] {
  return encodeBytes(new TextEncoder().encode(data));
}

/** xterm `onBinary` 每个 code unit 是一个原始 byte，不能再次做 UTF-8 编码。 */
export function encodeTerminalBinaryInput(data: string): string[] {
  const bytes = new Uint8Array(data.length);
  for (let index = 0; index < data.length; index += 1) bytes[index] = data.charCodeAt(index) & 0xff;
  return encodeBytes(bytes);
}

function encodeBytes(bytes: Uint8Array): string[] {
  const chunks: string[] = [];
  for (let offset = 0; offset < bytes.byteLength; offset += MAX_TERMINAL_CHUNK_BYTES) {
    chunks.push(bytesToBase64(bytes.subarray(offset, offset + MAX_TERMINAL_CHUNK_BYTES)));
  }
  return chunks;
}

export function clampTerminalDimension(value: number): number {
  if (!Number.isFinite(value)) return TERMINAL_DIM_MIN;
  return Math.min(TERMINAL_DIM_MAX, Math.max(TERMINAL_DIM_MIN, Math.trunc(value)));
}

export function decodeTerminalFrame(frame: Record<string, unknown>): TerminalDownstreamFrame | null {
  switch (frame.t) {
    case 'terminal_opened':
      return validId(frame.requestId)
        && validId(frame.terminalId)
        && typeof frame.cwd === 'string'
        && frame.cwd.length > 0
        && validDimension(frame.cols)
        && validDimension(frame.rows)
        ? frame as unknown as TerminalOpenedFrame
        : null;
    case 'terminal_output':
      return validId(frame.terminalId)
        && positiveSafeInteger(frame.seq)
        && strictBase64(frame.data)
        && decodedBase64Length(frame.data as string) <= MAX_TERMINAL_CHUNK_BYTES
        ? frame as unknown as TerminalOutputFrame
        : null;
    case 'terminal_exit':
      return validId(frame.terminalId)
        && optionalInteger(frame.exitCode)
        && optionalString(frame.signal)
        ? omitUndefined(frame, ['exitCode', 'signal']) as unknown as TerminalExitFrame
        : null;
    case 'terminal_error':
      return optionalId(frame.requestId)
        && optionalId(frame.terminalId)
        && typeof frame.code === 'string'
        && TERMINAL_ERROR_CODES.has(frame.code)
        && typeof frame.message === 'string'
        && typeof frame.retryable === 'boolean'
        ? omitUndefined(frame, ['requestId', 'terminalId']) as unknown as TerminalErrorFrame
        : null;
    default:
      return null;
  }
}

function validId(value: unknown): value is string {
  return typeof value === 'string'
    && value.length > 0
    && new TextEncoder().encode(value).byteLength <= MAX_TERMINAL_ID_BYTES;
}

const optionalId = (value: unknown): value is string | undefined => value == null || validId(value);
const optionalString = (value: unknown): value is string | undefined => value == null || typeof value === 'string';
const optionalInteger = (value: unknown): value is number | undefined => value == null || Number.isSafeInteger(value);
const validDimension = (value: unknown): value is number => Number.isSafeInteger(value)
  && Number(value) >= TERMINAL_DIM_MIN
  && Number(value) <= TERMINAL_DIM_MAX;
const positiveSafeInteger = (value: unknown): value is number => Number.isSafeInteger(value) && Number(value) > 0;

function strictBase64(value: unknown): value is string {
  if (typeof value !== 'string'
    || value.length > MAX_TERMINAL_BASE64_BYTES
    || value.length % 4 !== 0
    || !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(value)) return false;
  try {
    atob(value);
    return true;
  } catch {
    return false;
  }
}

function decodedBase64Length(value: string): number {
  if (!value) return 0;
  const padding = value.endsWith('==') ? 2 : value.endsWith('=') ? 1 : 0;
  return (value.length / 4) * 3 - padding;
}

function omitUndefined(frame: Record<string, unknown>, keys: string[]): Record<string, unknown> {
  const normalized = { ...frame };
  for (const key of keys) {
    if (normalized[key] == null) delete normalized[key];
  }
  return normalized;
}

File: proto/src/terminal.rs
//! 浏览器终端 PTY 线协议（独立于 ACP/Yjs/resource）。
//!
//! 原始字节经标准 base64 承载；`data` 解码后长度不得超过
//! [`MAX_TERMINAL_CHUNK_BYTES`]。

use base64::engine::general_purpose::STANDARD;
use base64::Engine;
use serde::{Deserialize, Serialize};

/// instance hello `caps.terminals.protocolVersion` 须为 1。
pub const TERMINAL_PROTOCOL_VERSION: u32 = 1;
/// 单次 input/output 原始字节上限。
pub const MAX_TERMINAL_CHUNK_BYTES: usize = 16 * 1024;
const MAX_TERMINAL_BASE64_BYTES: usize = MAX_TERMINAL_CHUNK_BYTES.div_ceil(3) * 4;
/// 终端行列下限（含）。
pub const TERMINAL_DIM_MIN: u16 = 2;
/// 终端行列上限（含）。
pub const TERMINAL_DIM_MAX: u16 = 500;
/// `requestId` / `projectId` / `terminalId` 最大 UTF-8 字节长度。
pub const MAX_TERMINAL_ID_BYTES: usize = 128;
/// 每个 browser principal 同时打开的终端数上限。
pub const MAX_TERMINALS_PER_PRINCIPAL: usize = 4;
/// instance 侧 PTY 会话总数上限。
pub const MAX_TERMINALS_PER_INSTANCE_HOST: usize = 16;

/// 稳定终端错误码（映射到 `terminal_error.code` 字符串）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum TerminalErrorCode {
    Forbidden,
    InvalidRequest,
    Unavailable,
    DeliveryUnknown,
    LimitExceeded,
}

impl TerminalErrorCode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Forbidden => "forbidden",
            Self::InvalidRequest => "invalid-request",
            Self::Unavailable => "unavailable",
            Self::DeliveryUnknown => "delivery-unknown",
            Self::LimitExceeded => "limit-exceeded",
        }
    }
}

// ---------------------------------------------------------------------------
// browser ↔ server
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOpen {
    pub request_id: String,
    pub project_id: String,
    pub cols: u16,
    pub rows: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalInput {
    pub terminal_id: String,
    pub seq: u64,
    pub data: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalResize {
    pub terminal_id: String,
    pub cols: u16,
    pub rows: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalClose {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub terminal_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOpened {
    pub request_id: String,
    pub terminal_id: String,
    pub cwd: String,
    pub cols: u16,
    pub rows: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOutput {
    pub terminal_id: String,
    pub seq: u64,
    pub data: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalExit {
    pub terminal_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exit_code: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signal: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TerminalError {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub terminal_id: Option<String>,
    pub code: TerminalErrorCode,
    pub message: String,
    pub retryable: bool,
}

// ---------------------------------------------------------------------------
// server ↔ instance
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalOpen {
    pub request_id: String,
    pub terminal_id: String,
    pub cwd: String,
    pub cols: u16,
    pub rows: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalInput {
    pub terminal_id: String,
    pub seq: u64,
    pub data: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalResize {
    pub terminal_id: String,
    pub cols: u16,
    pub rows: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalClose {
    pub terminal_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalOpened {
    pub request_id: String,
    pub terminal_id: String,
    pub ok: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cols: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rows: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalOutput {
    pub terminal_id: String,
    pub seq: u64,
    pub data: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstanceTerminalExit {
    pub terminal_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub exit_code: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signal: Option<String>,
}

/// 校验协议 id 字段：非空且 UTF-8 字节长度有界。
pub fn validate_terminal_id(id: &str, field: &str) -> Result<(), String> {
    if id.is_empty() {
        return Err(format!("{field} must not be empty"));
    }
    if id.len() > MAX_TERMINAL_ID_BYTES {
        return Err(format!("{field} exceeds max length"));
    }
    Ok(())
}

/// 校验终端尺寸。
pub fn validate_terminal_dims(cols: u16, rows: u16) -> Result<(), String> {
    if !(TERMINAL_DIM_MIN..=TERMINAL_DIM_MAX).contains(&cols)
        || !(TERMINAL_DIM_MIN..=TERMINAL_DIM_MAX).contains(&rows)
    {
        return Err("cols/rows out of range".into());
    }
    Ok(())
}

/// 严格 base64 解码并检查原始字节上限。
pub fn decode_terminal_chunk(data: &str) -> Result<Vec<u8>, String> {
    if data.is_empty() {
        return Ok(Vec::new());
    }
    if data.len() > MAX_TERMINAL_BASE64_BYTES {
        return Err("chunk too large".into());
    }
    let decoded = STANDARD
        .decode(data.as_bytes())
        .map_err(|_| "invalid base64".to_string())?;
    if decoded.len() > MAX_TERMINAL_CHUNK_BYTES {
        return Err("chunk too large".into());
    }
    Ok(decoded)
}

/// 编码输出块（调用方须保证 `raw.len() <= MAX_TERMINAL_CHUNK_BYTES`）。
pub fn encode_terminal_chunk(raw: &[u8]) -> String {
    STANDARD.encode(raw)
}

#[cfg(test)]
#[path = "terminal_test.rs"]
mod terminal_test;
